src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java

```diff
@@ -0,0 +1,533 @@
+                .findFirstToken(TokenTypes.SLIST)
+                .findFirstToken(TokenTypes.VARIABLE_DEF)
+                .getNextSibling()
+                .getNextSibling();
+        final DetailAST[] expected = {realVariableDefNode};
+        assertArrayEquals(expected, actual);
+    }
+
+    @Test
+    public void testParent() throws Exception {
+        final String xpath = "(//VARIABLE_DEF)[1]/..";
+        final DetailAST rootAst = getRootAst("InputXpathMapperAst.java");
+        final RootNode rootNode = new RootNode(rootAst);
+        final DetailAST[] actual = convertToDetailAstArray(getXpathItems(xpath, rootNode));
+        final DetailAST realVariableDefNode = getSiblingByType(rootAst, TokenTypes.CLASS_DEF)
+                .findFirstToken(TokenTypes.OBJBLOCK)
+                .findFirstToken(TokenTypes.METHOD_DEF)
+                .findFirstToken(TokenTypes.SLIST);
+        final DetailAST[] expected = {realVariableDefNode};
+        assertArrayEquals(expected, actual);
+    }
+
+    @Test
+    public void testCurlyBrackets() throws Exception {
+        final String xpath = "(//RCURLY)[2]";
+        final DetailAST rootAst = getRootAst("InputXpathMapperAst.java");
+        final RootNode rootNode = new RootNode(rootAst);
+        final DetailAST[] actual = convertToDetailAstArray(getXpathItems(xpath, rootNode));
+        final DetailAST realVariableDefNode = getSiblingByType(rootAst, TokenTypes.CLASS_DEF)
+                .findFirstToken(TokenTypes.OBJBLOCK)
+                .findFirstToken(TokenTypes.METHOD_DEF)
+                .findFirstToken(TokenTypes.SLIST)
+                .findFirstToken(TokenTypes.RCURLY);
+        final DetailAST[] expected = {realVariableDefNode};
+        assertArrayEquals(expected, actual);
+    }
+
+    @Test
+    public void testAnd() throws Exception {
+        final String xpath = "//CLASS_DEF | //METHOD_DEF";
+        final DetailAST rootAst = getRootAst("InputXpathMapperAst.java");
+        final RootNode rootNode = new RootNode(rootAst);
+        final DetailAST[] actual = convertToDetailAstArray(getXpathItems(xpath, rootNode));
+        final DetailAST realClassDefNode = getSiblingByType(rootAst, TokenTypes.CLASS_DEF);
+        final DetailAST realMethodDefNode1 = realClassDefNode
+                .findFirstToken(TokenTypes.OBJBLOCK)
+                .findFirstToken(TokenTypes.METHOD_DEF);
+        final DetailAST realMethodDefNode2 = realMethodDefNode1.getNextSibling();
+        final DetailAST[] expected = {realClassDefNode, realMethodDefNode1, realMethodDefNode2};
+        assertArrayEquals(expected, actual);
+    }
+
+    @Test
+    public void testQueryAllElementsWithAttribute() throws Exception {
+        final String xpath = "//*[@text]";
+        final DetailAST rootAst = getRootAst("InputXpathMapperAst.java");
+        final RootNode rootNode = new RootNode(rootAst);
+        final List<Item> nodes = getXpathItems(xpath, rootNode);
+        assertEquals(17, nodes.size());
```

Comment:
@timurt , please provide comment to each "assertXXXX" as first argument, it will be enforced by TC very soon.
@Nimfadora , please speedup fixes for messages, looks like other students try to give you more work.